src/osd/PeeringState.cc

```diff
@@ -986,6 +1034,113 @@ bool PeeringState::set_force_backfill(bool b)
   return did;
 }
 
+void PeeringState::recalc_readable_until(ceph::time_detail::signedspan now)
+{
+  ceph::signedspan interval = pool.get_readable_interval();
+
+  if (is_primary()) {
+    if (hb_stamps.empty()) {
+      // we are all alone
+      readable_until = now + interval;
+      readable_until_ub = readable_until;
+    } else {
+      ceph::signedspan max_ls = ceph::signedspan::zero();  // last_sent
+      ceph::signedspan max_lap = ceph::signedspan::zero(); // last_acked_ping
+      for (auto p = hb_stamps.begin(); p != hb_stamps.end(); ++p) {
+	ceph::signedspan ls, lap;
+	(*p)->sample_primary(&ls, &lap);
+	if (ls > max_ls) {
+	  max_ls = ls;
+	}
+	if (lap > max_lap) {
+	  max_lap = lap;
+	};
+      }
+      if (max_ls != ceph::signedspan::zero()) {
+	// we sent a ping and therefore replicas may be readable; bound based
+	// on our last send time
+	readable_until_ub = max_ls + interval;
+	if (max_lap != ceph::signedspan::zero()) {
+	  // we've received replies; make our readability something
+	  // <= the replicas' upper bounds (i.e., base it on our ping
+	  // *send* time)
+	  readable_until = max_lap + interval;
+	} else {
+	  // we haven't received any replies; do not allow reads
+	  readable_until = ceph::signedspan::zero();
+	}
+      } else {
+	// we haven't sent a ping; nobody is readable
+	readable_until = ceph::signedspan::zero();
+	readable_until_ub = ceph::signedspan::zero();
```

Comment:
Could we bootstrap readable_until and readable_until_ub during the info/notify exchange at the end of activation to avoid having to wait for a ping?

Reply:
I suppose it's likely that a ping will have been sent out before that point anyway.